### ISSUE TITLE
Direction of this project

### ISSUE THREAD

**Issue by macmania** [sjwhitworth/golearn#7]
I was wondering if it would be better to focus on a particular class in machine learning that is computationally intensive - dimensionality reduction, neural networks, fourier transform, etc rather than re-writing the algorithms. My take is if we just use one of the multi-translator compiler - python -> go or c++ -> go, we might be able to save some time. The team can just do some tests.

**Comment by anzellai** [comment]
I love the presentation and docs on this project, I think this is going to the right direction.
Here I would like to suggest going forward, we may implement a simple "WEB/API" package which can be used as a drop-in replacement to some external service, or simply a showcase for real life usage.

This is a package that may contain a few simple API endpoints and being very simple to understand and use (inspired by [Seldon project](http://docs.seldon.io/pluggable-prediction-algorithms.html), for example:

- /event/ endpoint to consume data input (general or with model schema definition).
- /predict/ endpoint to basically output prediction of what have been trained/consume so far

This can attract wider audience with programming/web background who are seeking simple prediction/recommendation solution without strong ML background (myself being one of them).

Currently although this project is great in golang, if I were to deploy something related to ML in web technology like Python/Java I'm going to build with `statsmodels` or opt for external service because they are easy to use and understand, but given the awesomeness of golang concurrency and much better latency for API development, this project can also provide a great alternative for those people how to make a simple use case.

What do you think @sjwhitworth and everyone? Although I'm only beginner of Golang, I would like to contribute to this great project to as much extent as I can.

**Comment by nickpoorman** [comment]
I think it's a good idea so people can get an understanding of how to use this project. I do think such a thing should be a separate project repo though.

**Comment by anzellai** [comment]
@nickpoorman I don't disagree, my suggestion is simply a way forward how to attract wider audience, and follow this project spirit, being simple to use and understand.

Let see how people think about this idea and we may consider how to implement this.

**Comment by Sentimentron** [comment]
I've spent some time thinking about this and I think it's a good idea. We need to improve our APIs to support various streaming/low-volume retraining and prediction events. At the moment, things in _base_ assume that things will generally a fixed size, I think it's time to change that assumption.

**Comment by nickpoorman** [comment]
@Sentimentron +1

**Comment by anzellai** [comment]
+1 perhaps we can create a **resources** package simply wrapped with an interface of a user defined model to provide instant API endpoints something like: 

/api/*{resources}*/data/                     [POST]
/api/*{resources}*/event/                   [POST]
/api/*{resources}*/predict/**               [GET]
/api/*{resources}*/report/*{method}*/ [GET]

We can also keep it really simple and leave all authentication stuff for user to implement.

**Comment by savorywatt** [comment]
I specifically wrote my own knn,  and k means clustering algorthims so I
could work with them in a service because golearn had a more data analysis
and very static approach to modeling and training. Just feedback about
this. At the time (a year ago) it made sense. Things may have changed. It
might make sense to think about how the library could be used in a service
and have an example of using things in a service. I think a full ML api
that adapts and retrains is a little beyond a core libraries scope.

My 2c

**Comment by anzellai** [comment]
@savorywatt I agree with your points and probably it's really time to start a separate repo to implement a full ML api. I also think some database supports would be a good idea.

For everyone else, may I ask who would like to drive this forward? I would like to help (or be part of it) and make this happen.

**Comment by chrmang** [comment]
This issue is open for more that 6 years now. And it's not about something, that can be fixed in code. 
Can it be closed?
